pennylane/devices/device_api.py

```diff
@@ -199,10 +201,72 @@ def wires(self) -> Wires:
         """
         return self._wires
 
+    def setup_execution_config(
+        self,
+        circuits: QuantumTape_or_Batch,
+        execution_config: ExecutionConfig = DefaultExecutionConfig,
+    ) -> ExecutionConfig:
+        """Fill in ambiguous entries for the execution config.
+
+        Args:
+            circuits (Union[QuantumTape, Sequence[QuantumTape]]): the quantum circuits to be executed
+            execution_config (ExecutionConfig): a datastructure with additional information required for execution
+
+        Returns:
+            ExecutionConfig: a new execution config with fully specified choices.
+
+        The execution config serves to communicate both how the device wants to be used by PennyLane, and
+        how the device will perform its computation. This method gives the device an opportunity to specify how
+        it wants to be used and add information on its configuration.
+
+        If the device is capable of handling the derivative, it should set ``use_device_gradient`` to be ``True``.
+
+        >>> config = ExecutionConfig(use_device_gradient=None, gradient_method="device")
+        >>> new_config = dev.setup_execution_config(circuit, config)
+        >>> new_config.use_device_gradient
+        True
+
+        For example, the initial ``execution_config`` may have a ``gradient_method`` of ``"best"``.
+        The device can then specify what is its preferred gradient method.
+
+        >>> config = ExecutionConfig(gradient_method="best", grad_on_execution=None)
+        >>> circuit = qml.tape.QuantumScript([], [qml.state()])
+        >>> new_config = dev.setup_execution_config(circuit, config)
+        >>> new_config.gradient_method
+        'backprop'
+
+        Or for a device that prefers adjoint:
+
+        >>> new_config = adjoint_dev.setup_execution_config(circuit, config)
+        >>> new_config.gradient_method
+        'adjoint'
+        >>> new_config.grad_on_execution
+        True
+
+        The device should also fill in ``device_options`` with any device specific defaults:
+
+        >>> new_config.device_options
+        {'max_workers': None, 'rng': Generator(PCG64) at 0x17D9C67A0, 'prng_key': None}
+
+        """
+        updated_values = {}
+        if (
+            execution_config.gradient_method in {"best", "device"}
+            and type(self).compute_derivatives != Device.compute_derivatives
+        ):
```

Comment:
do we want this function to raise errors? what if the user asks for "device" but doesn't override `compute_derivates`?